### PR TITLE
Email service

### PR DESCRIPTION
**Pull request by branh0821**
Note: For local tests you need to configure your environmental variables to an API key generated from SendGrid, as well as change the placeholder sender address to the one tied to the API key. Otherwise the Rust API panics because there is a discrepancy between the API key and the address value it's supposed to be tied to.

### CODE REVIEW

File: packages/rust/server/src/routes/user.rs [hunk 1]
```diff
@@ -93,10 +94,16 @@ async fn register(
     })?;
 
     email_service
-        .send(&format!(
-            "to: {} | body: token: {}",
-            confirmation.email, confirmation.token
-        ))
+        .send(
+            &format!(
+                "to: {} | body: token: {}",
```
Comment (jbis9051) [inline packages/rust/server/src/routes/user.rs]: nit: change this to something a bit more usable

File: packages/rust/server/src/routes/user.rs [hunk 2]
```diff
@@ -223,10 +230,13 @@ async fn forgot(
     forgot.create(&db.0).await.map_err(map_sqlx_err)?;
 
     email_service
-        .send(&format!(
-            "to: {} | body: token: {}",
-            payload.email, forgot.token
-        ))
+        .send(
+            &format!("to: {} | body: token: {}", payload.email, forgot.token),
```
Comment (jbis9051) [inline packages/rust/server/src/routes/user.rs]: same here

File: packages/rust/server/src/services/email.rs [hunk 3]
```diff
@@ -1,10 +1,95 @@
+use crate::config::CONFIG;
+use sendgrid::SGClient;
+use sendgrid::SendgridError;
+use sendgrid::{Destination, Mail};
+
 #[derive(Default, Clone, Copy)]
 pub struct EmailService {}
+pub struct MockEmailService {}
+
+pub trait Email_ {
```
Comment (jbis9051) [inline packages/rust/server/src/services/email.rs]: trait should be called EmailService, you should have two structs SendgridEmailService and PrinterEmailService or something like that. Trait also needs to be implemented for all routes.

File: packages/rust/server/src/services/email.rs [hunk 4]
```diff
@@ -1,10 +1,95 @@
...
+        _recipient: &Vec<(String, String)>,
+        _hasHTML: bool,
+        _html_content: &str,
+    ) -> Result<(), SendgridError>;
+}
+
+impl Email_ for EmailService {
+    fn send(
+        &self,
+        _email: &str,
```
Comment (jbis9051) [inline packages/rust/server/src/services/email.rs]: underscore pattern is only for unused variables

File: packages/rust/server/src/services/email.rs [hunk 5]
```diff
@@ -1,10 +1,95 @@
...
+impl Email_ for MockEmailService {
+    fn send(
+        &self,
+        _email: &str,
+        _subject: &str,
+        _recipients: &Vec<(String, String)>,
+        _hasHTML: bool,
+        _html_content: &str,
+    ) -> Result<(), SendgridError> {
+        let _api_key = &CONFIG.api_key_check;
```
Comment (jbis9051) [inline packages/rust/server/src/services/email.rs]: why do we need an api key for a mockemailservice?

File: packages/rust/server/tests/email_service_test.rs [hunk 6]
```diff
@@ -0,0 +1,32 @@
+#[cfg(test)]
```
Comment (jbis9051) [inline packages/rust/server/tests/email_service_test.rs]: No tests are needed for the email service

File: packages/rust/server/src/config.rs [hunk 7]
```diff
@@ -4,9 +4,13 @@ use std::env;
 pub struct Config {
     pub listen_addr: String,
     pub db_url: String,
+    pub api_key_check: String,
+    pub sender_email: String,
```
Comment (jbis9051) [inline packages/rust/server/src/config.rs:8]: not: capitalize the env var

File: packages/rust/server/src/config.rs [hunk 8]
```diff
@@ -4,9 +4,13 @@ use std::env;
...
 }
 
 pub static CONFIG: Lazy<Config> = Lazy::new(|| Config {
     listen_addr: env::var("LISTEN_ADDR").unwrap(),
     db_url: env::var("DB_URL").unwrap(),
+    api_key_check: env::var("SENDGRID_API_KEY").unwrap(), // pull api key from env. variables
+    sender_email: env::var("sender_email").unwrap(),
```
Comment (jbis9051) [inline packages/rust/server/src/config.rs]: nit: capitalize the env var

File: packages/rust/server/src/services/email.rs [hunk 9]
```diff
@@ -1,10 +1,94 @@
...
 #[derive(Default, Clone, Copy)]
-pub struct EmailService {}
+pub struct PrinterEmailService {}
+
+pub trait EmailService {
+    fn send(
+        &self,
+        email: &str,
+        subject: &str,
+        recipient: &Vec<(String, String)>,
```
Comment (jbis9051) [inline packages/rust/server/src/services/email.rs]: what's the difference between email and recipient? and do we need a Vec or will a slice work?

File: packages/rust/server/src/services/email.rs [hunk 10]
```diff
@@ -1,10 +1,94 @@
...
+impl EmailService for PrinterEmailService {
+    fn send(
+        &self,
+        email: &str,
+        subjects: &str,
+        recipients: &Vec<(String, String)>,
+        hash_html: bool,
+        html_content: &str,
+    ) -> Result<(), SendgridError> {
+        let mut mail_info = Mail::new()
```
Comment (jbis9051) [inline packages/rust/server/src/services/email.rs]: can we simplify the print implementation? just print it?

File: packages/rust/server/src/services/email.rs (hunk 9 above)
Comment (branh0821) [reply packages/rust/server/src/services/email.rs]: This is an array of tuples/pairs that package a recipient's name and email address. I could make it a slice.

**Review by jbis9051 — APPROVED** [review]

**Pull request merged**